### PR TITLE
[UX] Remove warning for gcp check

### LINKED ISSUE #3060: [UX] `sky check` shows warning from GCP package
The following message shows during the `sky check`, which may related to some deprecation/update from google API. We should check the cause and fix them.

```console
$ sky check
Checking credentials to enable clouds for SkyPilot.
  Checking GCP...INFO:googleapiclient.discovery_cache:file_cache is only supported with oauth2client<4.0.0
  GCP: enabled                              
  AWS: enabled                              
  Azure: enabled
  ...
```

**Versions:**
``` console
$ gcloud --version
Google Cloud SDK 461.0.0
beta 2024.01.22
bundled-python3-unix 3.11.6
core 2024.01.22
Updates are available for some Google Cloud CLI components.  To install them,
please run:
  $ gcloud components update

$ pip list | grep google
google-api-core                         2.14.0
google-api-python-client                2.107.0
google-auth                             2.23.4
google-auth-httplib2                    0.1.1
google-cloud-core                       2.3.3
google-cloud-dataproc                   5.8.0
google-cloud-storage                    2.13.0
google-crc32c                           1.5.0
google-resumable-media                  2.6.0
googleapis-common-protos                1.61.0
grpc-google-iam-v1                      0.13.0
```

### PR DESCRIPTION
**Pull request by Michaelvll**
Fixes #3060

This also adds the detailed error for GCP operation in the exception raised to make it easier to debug.

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
  - [x] `sky check gcp`
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`

### CODE REVIEW

File: sky/provision/gcp/instance_utils.py [hunk 1]
```diff
@@ -100,19 +100,20 @@ def _generate_node_name(cluster_name: str, node_suffix: str,
     return node_name
 
 
-def _log_errors(errors: List[Dict[str, str]], e: Any,
-                zone: Optional[str]) -> None:
+def _format_message_from_errors(errors: List[Dict[str, str]], e: Any,
+                                zone: Optional[str]) -> str:
     """Format errors into a string."""
```
Comment (cblmemo) [inline sky/provision/gcp/instance_utils.py:105]:
```suggestion
def _format_message_from_errors_and_log(errors: List[Dict[str, str]], e: Any,
                                zone: Optional[str]) -> str:
    """Format errors into a string and log them to the console."""
```
to better align the actual behaviour of this function?

File: sky/clouds/gcp.py [hunk 2]
```diff
@@ -742,7 +742,8 @@ def check_credentials(cls) -> Tuple[bool, Optional[str]]:
         credentials, project = google.auth.default()
         crm = googleapiclient.discovery.build('cloudresourcemanager',
                                               'v1',
-                                              credentials=credentials)
+                                              credentials=credentials,
+                                              cache_discovery=False)
```
Comment (cblmemo) [inline sky/clouds/gcp.py:746]: Should we benchmark the performance degrade and make sure it is tolerable? I think it should be marginal though

File: sky/provision/gcp/instance_utils.py [hunk 3]
```diff
@@ -462,8 +464,9 @@ def call_operation(fn, timeout: int):
                 'message': f'Timeout waiting for operation {operation["name"]}',
                 'domain': 'wait_for_operation'
             }]
-            _log_errors(errors, None, zone)
+            msg = _format_message_from_errors(errors, None, zone)
             error = common.ProvisionerError('Operation timed out')
+            setattr(error, 'detailed_reason', msg)
```
Comment (cblmemo) [inline sky/provision/gcp/instance_utils.py:469]: Should we add an attribute for the class and assign here, instead of `setattr`?

**Review by cblmemo — COMMENTED** [review]
Thanks for the fix @Michaelvll ! Left some comments :))

File: sky/provision/gcp/instance_utils.py (hunk 3 above)
Comment (Michaelvll) [reply sky/provision/gcp/instance_utils.py:471]: This is only used for the usage collection, which uses `hasattr` to check `detailed_reason`. Maybe we can keep it this way?

File: sky/clouds/gcp.py (hunk 2 above)
Comment (Michaelvll) [reply sky/clouds/gcp.py]: Yep, the performance is similar. Tested with:
```
multitime -n 5 python -c "from sky.adaptors import gcp; gcp.build('cloudresourcemanager', 'v1', credentials=None, cache_discovery=False); gcp.build('cloudresourcemanager', 'v1', credentials=None, cache_discovery=False)"
===> multitime results
1: python -c "from sky.adaptors import gcp; gcp.build('cloudresourcemanager', 'v1', credentials=None, cache_discovery=False); gcp.build('cloudresourcemanager', 'v1', credentials=None, cache_discovery=False)"
            Mean        Std.Dev.    Min         Median      Max
real        1.174       0.009       1.165       1.174       1.188       
user        1.210       0.025       1.170       1.214       1.245       
sys         0.243       0.021       0.220       0.238       0.278
```

```
multitime -n 5 python -c "from sky.adaptors import gcp; gcp.build('cloudresourcemanager', 'v1', credentials=None); gcp.build('cloudresourcemanager', 'v1', credentials=None)"
INFO:googleapiclient.discovery_cache:file_cache is only supported with oauth2client<4.0.0
INFO:googleapiclient.discovery_cache:file_cache is only supported with oauth2client<4.0.0
INFO:googleapiclient.discovery_cache:file_cache is only supported with oauth2client<4.0.0
INFO:googleapiclient.discovery_cache:file_cache is only supported with oauth2client<4.0.0
INFO:googleapiclient.discovery_cache:file_cache is only supported with oauth2client<4.0.0
===> multitime results
1: python -c "from sky.adaptors import gcp; gcp.build('cloudresourcemanager', 'v1', credentials=None); gcp.build('cloudresourcemanager', 'v1', credentials=None)"
            Mean        Std.Dev.    Min         Median      Max
real        1.241       0.028       1.211       1.230       1.293       
user        1.262       0.038       1.220       1.269       1.323       
sys         0.262       0.019       0.237       0.258       0.294 
```

File: sky/provision/gcp/instance_utils.py (hunk 3 above)
Comment (cblmemo) [reply sky/provision/gcp/instance_utils.py:471]: Oh ic. LGTM!

**Review by cblmemo — APPROVED** [review]

**Pull request merged**